modules/annex/annex_archive.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package annex
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"code.gitea.io/gitea/modules/git"
+)
+
+func repoArchiveTargz(repo *git.Repository, target io.Writer, prefix, commitID string) error {
+	// create a plain git tar archive
+	var tarB bytes.Buffer
+	err := repo.CreateArchive(repo.Ctx, git.TAR, &tarB, prefix != "", commitID)
```

Comment:
As far as I can tell, this function is pretty efficient: it reads off the `git archive` stdout into `tarB` only as needed and then immediately turns around and writes that back out again. And you never need to double-back or buffer to disk space.

So I wonder if maybe the trick to getting this fully efficient is just this: replace `bytes.Buffer` with a type that truncates itself as soon as `Read()` is called on it? [`io.Pipe()`](https://pkg.go.dev/io#Pipe) should do that?


```suggestion
	tarB, tarBinput := io.Pipe()
	err := repo.CreateArchive(repo.Ctx, git.TAR, &tarBinput, prefix != "", commitID)
```


But I'm just musing, like we agreed this can be done in a follow-up.

Reply:
This would deadlock since `repo.CreateArchive` blocks on writing to the pipe until we read from it, which does not happen in this single threaded scenario. We would have to wrap `CreateArchive` in a go routine to do that. You can actually see how that would look in https://github.com/neuropoly/gitea/blob/864506d027b8f1154c39c7efd1982869640480cc/services/repository/archiver/archiver.go#L176, from where `repo.CreateArchive` and the annex implementation get called. Error handling becomes a lot more verbose due to requiring a channel and what not, so I decided against this, but I actually had it implemented this way at some point. It is only really feasible to do this "streaming" for tar anyway, since the zip reader requires an `io.ReaderAt` and the archive size beforehand. Since the archives created by `repo.CreateArchive` are mostly small anyway (I'd expect sub-Gigabyte in almost all cases) I think it is not necessary to optimize this. There is an approximate upper-bound in the memory usage this can have in `(plain-git size of largest repository) * (number of workers in the repo-archive queue)`, the latter being:
> MAX_WORKERS: (dynamic): Maximum number of worker go-routines for the queue. Default value is "CpuNum/2" clipped to between 1 and 10.

What I actually meant is that the download in the browser of the user requesting the archive only starts after the server has completely generated the archive first. That can mean quite a long wait until anything happens. This is due to the archive http endpoint waiting for the archive request to be finished in https://github.com/neuropoly/gitea/blob/864506d027b8f1154c39c7efd1982869640480cc/routers/api/v1/repo/file.go#L310 and the used [http.ServeContent](https://pkg.go.dev/net/http#ServeContent) requiring an `io.ReadSeeker`.